### DISCUSSION TITLE
How Can i Create a Scrollable list of the Number's using the egui like in the image but 1 Column.

### DISCUSSION THREAD

**Discussion by SohamTilekar** [emilk/egui#4903] [Q&A]
How Can i Create a Scrollable list of the Number's using the egui & eframe like in the image but 1 Column.
```ps
00  data |
02  data |
02  data |
```
data is the 32 bit hex char.
and `|` is the scrolling bar

The Data Is an long Array So it is Super Slow, & I reeded the the Doc.

![image](https://github.com/user-attachments/assets/b281f61e-7230-45c3-aec4-1d66214b4b25)

**Comment by SohamTilekar** [comment]
I trid the Things From demo But it didn't work because the array length is 2^24

**Comment by YgorSouza** [comment]
You mean something like this with the [show_rows](https://docs.rs/egui/0.28.1/egui/containers/scroll_area/struct.ScrollArea.html#method.show_rows) method?

```rs
struct MyApp {
    values: Vec<u32>,
}

impl Default for MyApp {
    fn default() -> Self {
        Self {
            values: (1..(1 << 23)).collect(),
        }
    }
}

impl eframe::App for MyApp {
    fn update(&mut self, ctx: &egui::Context, _frame: &mut eframe::Frame) {
        egui::CentralPanel::default().show(ctx, |ui| {
            let height = TextStyle::Body.resolve(ui.style()).size;
            ScrollArea::vertical().show_rows(ui, height, self.values.len(), |ui, row_range| {
                ui.allocate_space([ui.available_width(), 0.0].into());
                for i in row_range {
                    let Some(value) = self.values.get(i) else {
                        continue;
                    };
                    ui.label(format!("{value:08x}"));
                }
            })
        });
    }
}
```

![image](https://github.com/user-attachments/assets/c7e125df-d171-4cc0-b0b2-3b27daaa5c11)


This seems to work fine up to 2^23 items, but with more than that you start having floating-point precision problems, since all the positions are stored in f32 which has 24 bits of precision.

**Reply by SohamTilekar** [reply]
Another Help, How Can I Change the Scrolling Wheal From The Code, Because I Wana to Add the Search Bar on the Top Which will take it to a Particular Adress.

**Reply by SohamTilekar** [reply]
When I try to Put Both Side By side then they Don't extend till end and if i Move 1 Then Both Move's.
I can Change its height using the `min_height` but both move if I Try to Move ever a Single One.
![image](https://github.com/user-attachments/assets/1c11fac1-2f20-41cd-aa9a-c1f9cab171b5)
Code:
```rust
ui.horizontal(|ui| {
    // Stack
    ui.vertical(|ui| {
        ui.label("Stack Contents:");
        let height = egui::TextStyle::Body.resolve(ui.style()).size;
        egui::ScrollArea::vertical().show_rows(
            ui,
            height,
            cpu.ram[0xFFF400..=0xFFFFFF].len(),
            |ui, row_range| {
                for i in row_range {
                    let Some(value) = cpu.ram.get(0xFFF400 + i) else {
                        continue;
                    };
                    ui.label(format!("SP-{}: 0x{:08x}", i * 4, value));
                }
            },
        );
    });
    // RAM
    ui.vertical(|ui| {
        ui.label("RAM Contents:");
        let height = egui::TextStyle::Body.resolve(ui.style()).size;
        egui::ScrollArea::vertical().show_rows(
            ui,
            height,
            cpu.ram.len(),
            |ui, row_range| {
                for i in row_range {
                    let Some(value) = cpu.ram.get(i) else {
                        continue;
                    };
                    ui.label(format!("{}: 0x{:08x}", i * 4, value));
                }
            },
        );
    });
});
```

**Reply by YgorSouza** [reply]
> Another Help, How Can I Change the Scrolling Wheal From The Code, Because I Wana to Add the Search Bar on the Top Which will take it to a Particular Adress.

You can use [vertical_scroll_offset](https://docs.rs/egui/0.28.1/egui/containers/scroll_area/struct.ScrollArea.html#method.vertical_scroll_offset) when the user clicks on a button or changes the search text. Here's an example:

```rs
struct MyApp {
    values: Vec<u32>,
    search: String,
}

impl Default for MyApp {
    fn default() -> Self {
        Self {
            values: (1..(1 << 23)).collect(),
            search: String::new(),
        }
    }
}

impl eframe::App for MyApp {
    fn update(&mut self, ctx: &egui::Context, _frame: &mut eframe::Frame) {
        egui::CentralPanel::default().show(ctx, |ui| {
            let mut search = None;
            if ui.text_edit_singleline(&mut self.search).changed() {
                search = u32::from_str_radix(&self.search, 16).ok();
            }
            let height = TextStyle::Body.resolve(ui.style()).size;
            let mut scroll_area = ScrollArea::vertical();
            if let Some(index) = search.and_then(|s| self.values.iter().position(|v| v == &s)) {
                let height_with_spacing = height + ui.spacing().item_spacing.y;
                let offset = height_with_spacing * index as f32;
                scroll_area = scroll_area.vertical_scroll_offset(offset);
            }
            scroll_area.show_rows(ui, height, self.values.len(), |ui, row_range| {
                ui.allocate_space([ui.available_width(), 0.0].into());
                for i in row_range {
                    let Some(value) = self.values.get(i) else {
                        continue;
                    };
                    ui.label(format!("{value:08x}"));
                }
            })
        });
    }
}
```

![image](https://github.com/user-attachments/assets/26caddcf-4b7f-4767-8dd1-958bc1386517)

**Reply by YgorSouza** [reply]
> When I try to Put Both Side By side then they Don't extend till end and if i Move 1 Then Both Move's.

If you have multiple scroll areas in the same UI, you have to give each of them a different `Id` using the [id_source](https://docs.rs/egui/0.28.1/egui/containers/scroll_area/struct.ScrollArea.html#method.id_source) method.

**Reply by SohamTilekar** [reply]
I Try to set the `.min_scrolled_height(ui.available_height())` but it did this.
![image](https://github.com/user-attachments/assets/af58950a-bbf8-49b2-90eb-5af37a02802a)

**Reply by YgorSouza** [reply]
Hard to know why this is happening without seeing the whole code. You can use the debug tools to see how much space is being allocated for each UI, and then adjust the layouts to do what you want.

```rs
ui.painter().debug_rect(ui.max_rect(), Color32::GREEN, "UI1");
ui.debug_paint_cursor();
```

**Reply by SohamTilekar** [reply]
@YgorSouza The Searcher DO not Work with More than 1 Input Character entered it only used the 1 character.

**Reply by YgorSouza** [reply]
> @YgorSouza The Searcher DO not Work with More than 1 Input Character entered it only used the 1 character.

In the example I gave you? It's working fine for me.

```rs
            let mut search = None;
            if ui.text_edit_singleline(&mut self.search).changed() {
                search = u32::from_str_radix(&self.search, 16).ok();
            }
```

This is going to parse whatever you type as a hex string, and put it in the number variable so you can search for it in the data array.
